pkg/apis/pipeline/v1alpha1/build_gcs_resource.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2019 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"flag"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	buildGCSFetcherImage = flag.String("gcs-fetcher-image", "gcr.io/cloud-builders/gcs-fetcher:latest",
+		"The container image containing our GCS fetcher binary.")
+	buildGCSUploaderImage = flag.String("gcs-uploader-image", "gcr.io/cloud-builders/gcs-uploader:latest",
+		"The container image containing our GCS uploader binary.")
+)
+
+// GCSArtifactType defines a type of GCS resource.
+type GCSArtifactType string
+
+const (
+	// GCSArchive indicates that resource should be fetched from a typical archive file.
+	GCSArchive GCSArtifactType = "Archive"
+
+	// GCSManifest indicates that resource should be fetched using a
+	// manifest-based protocol which enables incremental source upload.
+	GCSManifest GCSArtifactType = "Manifest"
+)
+
+// BuildGCSResource describes a resource in the form of an archive,
+// or a source manifest describing files to fetch.
+type BuildGCSResource struct {
```

Comment:
Instead of `BuildGCSResource` can we call this something more descriptive, like `GCSFetcherResource`? The difference seems to be that this type of resource uses the fetcher image

Reply:
Both GCSResource and BuildGCSResource  can fetch and upload the GCS resource.
Both implement the `PipelineResourceInterface`.
```
// PipelineResourceInterface interface to be implemented by different PipelineResource types
type PipelineResourceInterface interface {
	GetName() string
	GetType() PipelineResourceType
	GetParams() []Param
	Replacements() map[string]string
	GetDownloadContainerSpec() ([]corev1.Container, error)
	GetUploadContainerSpec() ([]corev1.Container, error)
	SetDestinationDirectory(string)
}
```

The only difference between 2 are:
1. They  use different images 
2. The BuildGCSResource supports `--type=Manifest` which supports incremental upload of sources from a client.

Reply:
Hence i feel, we should get rid of redundancy and just define one GCSResource.

Reply:
Another difference is Build GCS resource supports `Manifest` and `Archive` type only. Additionally when `Archive` type is used, GCS fetcher image unzips the contents and places them in destination dir. Whereas GCS storage resource supports all types and does not do anything downloaded content.

Reply:
> we should get rid of redundancy and just define one GCSResource.

Do we imagine a lot of people will want to use the Archive + Manifest functionality? I'm not familiar with how this came into existence, so in my mind these are like specialized, not frequently used features, and bare uploads and downloads to a gcs bucket seem more normal.

But maybe im wrong about that, and the Archive + Manifest functionality should be considered like key features? If so, then I agree with @tejal29 and having once Resource that can support all 3 modes seems like the way to do.

Wnat do you think @shashwathi  - we could have have the existing GCS resource support all 3?

Reply:
The archive and Manifest functionality is for supporting the use case where source code is on GCS either as a tar or as whole repo.
It is used in Build for fetching source. I don't think we sync changes if any made by build back back to source. But if we do want to support that like in @dlorenc  Document, we need to support uploading via Manifest